File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/collectors/EmulatorCollector.java
package cn.tongdun.mobrisk.core.collectors;

import android.annotation.SuppressLint;
import android.os.Build;

import java.io.File;

/**
 * @description: detect Emulator
 * @author: wuzuchang
 * @date: 2023/4/25
 */
public class EmulatorCollector {

    private static final String[] QEMU_FILES = {
            "/dev/socket/genyd",
            "/dev/socket/baseband_genyd",
            "/dev/socket/qemud",
            "/dev/qemu_pipe",
            "/system/bin/nox-prop",
            "/system/bin/ttVM-prop",
            "/system/bin/droid4x-prop",
            "/data/.bluestacks.prop",
            "/system/bin/androVM-prop",
            "/system/bin/microvirt-prop",
            "/system/lib/libdroid4x.so",
            "/system/bin/windroyed",
            "/system/bin/microvirtd"
    };

    public boolean detectEmulator() {
        boolean isEmulator = checkBuildInfo();
        if (!isEmulator){
            isEmulator = checkFiles();
        }
        return isEmulator;
    }

    @SuppressLint("HardwareIds")
    private boolean checkBuildInfo() {
        return "nox".equals(Build.DEVICE)
                || "vbox86p".equals(Build.DEVICE)
                || "vbox86tp".equals(Build.DEVICE)
                || "appplayer".equals(Build.DEVICE)
                || "droid4x".equals(Build.DEVICE)
                || "vbox".equals(Build.DEVICE)
                || "virtual".equals(Build.DEVICE)
                || "andywin".equals(Build.DEVICE)
                || "andyosx".equals(Build.DEVICE)
                || "generic".equals(Build.DEVICE)
                || "generic_x86".equals(Build.DEVICE)
                || Build.DEVICE.toLowerCase().contains("mumu")
                || Build.DEVICE.toLowerCase().contains("zerofltezc")
                || Build.FINGERPRINT.startsWith("generic")
                || "vmos".equals(Build.MODEL)
                || "duos".equals(Build.MODEL)
                || "amiduos".equals(Build.MODEL)
                || "noxw".equals(Build.MODEL)
                || "genymotion".equals(Build.MODEL)
                || "bluestacks".equals(Build.MODEL)
                || "tiantian".equals(Build.MODEL)
                || "windroy".equals(Build.MODEL)
                || Build.MODEL.contains("google_sdk")
                || Build.MODEL.toLowerCase().contains("droid4x")
                || Build.MODEL.contains("Emulator")
                || Build.MODEL.contains("Android SDK built for x86")
                || Build.MODEL.contains("Subsystem for Android(TM)")
                || Build.MANUFACTURER.contains("Genymotion")
                || "goldfish".equals(Build.HARDWARE)
                || "vbox86".equals(Build.HARDWARE)
                || "sdk".equals(Build.PRODUCT)
                || "google_sdk".equals(Build.PRODUCT)
                || "sdk_x86".equals(Build.PRODUCT)
                || "vbox86p".equals(Build.PRODUCT)
                || "vbox86tp".equals(Build.PRODUCT)
                || "genymotion".equals(Build.PRODUCT)
                || "bluestacks".equals(Build.PRODUCT)
                || "droid4x".equals(Build.PRODUCT)
                || "ttvm_hdragon".equals(Build.PRODUCT)
                || "duos_native".equals(Build.PRODUCT)
                || "duos".equals(Build.PRODUCT)
                || "vbox".equals(Build.PRODUCT)
                || "android_x86".equals(Build.PRODUCT)
                || Build.PRODUCT.toLowerCase().contains("nox")
                || Build.BOARD.toLowerCase().contains("nox")
                || Build.BOOTLOADER.toLowerCase().contains("nox")
                || Build.HARDWARE.toLowerCase().contains("nox")
                || Build.SERIAL.toLowerCase().contains("nox")
                || Build.HOST.toLowerCase().startsWith("bliss-os")
                || (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic"));
    }

    private boolean checkFiles() {
        for (String filePtah : QEMU_FILES) {
            File qemu_file = new File(filePtah);
            if (qemu_file.exists()) {
                return true;
            }
        }
        return false;
    }

}

File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/collectors/EmulatorCollector.kt
package cn.tongdun.mobrisk.core.collectors

import android.os.Build
import cn.tongdun.mobrisk.core.collectors.EmulatorCollector
import java.io.File
import java.util.*

/**
 * @description:
 * @author: wuzuchang
 * @date: 2023/5/15
 */
interface EmulatorInterface {
    fun detectEmulator(): Boolean
}

class EmulatorCollector : EmulatorInterface {

    private val QEMU_FILES = listOf(
        "/dev/socket/genyd",
        "/dev/socket/baseband_genyd",
        "/dev/socket/qemud",
        "/dev/qemu_pipe",
        "/system/bin/nox-prop",
        "/system/bin/ttVM-prop",
        "/system/bin/droid4x-prop",
        "/data/.bluestacks.prop",
        "/system/bin/androVM-prop",
        "/system/bin/microvirt-prop",
        "/system/lib/libdroid4x.so",
        "/system/bin/windroyed",
        "system/lib/libnoxspeedup.so",
        "/system/bin/duosconfig",
        "/system/etc/xxzs_prop.sh",
        "/system/etc/mumu-configs/device-prop-configs/mumu.config",
        "/system/priv-app/ldAppStore",
        "system/bin/ldinit",
        "/system/app/AntStore",
        "vmos.prop",
        "fstab.titan",
        "x8.prop",
        "/system/lib/libc_malloc_debug_qemu.so"
    )

    private fun checkDevice(): Boolean = "nox" == Build.DEVICE
            || "vbox86p" == Build.DEVICE
            || "vbox86tp" == Build.DEVICE
            || "appplayer" == Build.DEVICE
            || "droid4x" == Build.DEVICE
            || "vbox" == Build.DEVICE
            || "virtual" == Build.DEVICE
            || "andywin" == Build.DEVICE
            || "andyosx" == Build.DEVICE
            || "generic" == Build.DEVICE
            || "generic_x86" == Build.DEVICE
            || "emu64a" == Build.DEVICE
            || Build.DEVICE.lowercase(Locale.getDefault()).contains("mumu")
            || Build.DEVICE.lowercase(Locale.getDefault()).contains("zerofltezc")

    private fun checkModel(): Boolean = "vmos" == Build.MODEL
            || "duos" == Build.MODEL
            || "amiduos" == Build.MODEL
            || "noxw" == Build.MODEL
            || "genymotion" == Build.MODEL
            || "bluestacks" == Build.MODEL
            || "tiantian" == Build.MODEL
            || "windroy" == Build.MODEL
            || Build.MODEL.contains("google_sdk")
            || Build.MODEL.lowercase(Locale.getDefault()).contains("droid4x")
            || Build.MODEL.contains("Emulator")
            || Build.MODEL.contains("Android SDK built for x86")
            || Build.MODEL.contains("Subsystem for Android(TM)")

    private fun checkFingerprint(): Boolean = Build.FINGERPRINT.startsWith("generic")
            || Build.FINGERPRINT.contains("vbox")

    private fun checkProduct(): Boolean = "sdk" == Build.PRODUCT
            || "google_sdk" == Build.PRODUCT
            || "sdk_x86" == Build.PRODUCT
            || "vbox86p" == Build.PRODUCT
            || "vbox86tp" == Build.PRODUCT
            || "genymotion" == Build.PRODUCT
            || "bluestacks" == Build.PRODUCT
            || "droid4x" == Build.PRODUCT
            || "ttvm_hdragon" == Build.PRODUCT
            || "duos_native" == Build.PRODUCT
            || "duos" == Build.PRODUCT
            || "vbox" == Build.PRODUCT
            || "android_x86" == Build.PRODUCT
            || Build.PRODUCT.lowercase(Locale.getDefault()).contains("nox")

    private fun checkHardware(): Boolean = "goldfish" == Build.HARDWARE
            || "vbox86" == Build.HARDWARE

    private fun checkOtherBuildInfo(): Boolean = Build.MANUFACTURER.contains("Genymotion")
            || Build.BOARD.lowercase(Locale.getDefault()).contains("nox")
            || Build.BOOTLOADER.lowercase(Locale.getDefault()).contains("nox")
            || Build.HARDWARE.lowercase(Locale.getDefault()).contains("nox")
            || Build.SERIAL.lowercase(Locale.getDefault()).contains("nox")
            || Build.HOST.lowercase(Locale.getDefault()).startsWith("bliss-os")
            || Build.HOST.lowercase(Locale.getDefault()).contentEquals("G1-SNIPER-B7")
            || Build.HOST.lowercase(Locale.getDefault()).contentEquals("Build2")
            || (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic"))

    override fun detectEmulator(): Boolean {
        var isEmulator = checkBuildInfo()
        if (!isEmulator) {
            isEmulator = checkFiles()
        }
        return isEmulator
    }

    private fun checkBuildInfo(): Boolean =
        checkDevice() ||
                checkModel() ||
                checkFingerprint() ||
                checkProduct() ||
                checkHardware() ||
                checkOtherBuildInfo()

    private fun checkFiles(): Boolean {
        for (filePtah in QEMU_FILES) {
            if (File(filePtah).exists()) {
                return true
            }
        }
        return false
    }
}